### PR TITLE
litdata backed tinystories

### LINKED ISSUE #995: Use litdata for TinyStories tokenization
Could you follow the OpenWebText data module for the implementation? This would streamline the data loading and so we don't have to reimplement the chunking / packed-dataset which we had before.

_Originally posted by @awaelchli in https://github.com/Lightning-AI/lit-gpt/pull/980#discussion_r1511381177_

### LINKED ISSUE #1085: Issues with TinyStories
I tried to pretrain with TinyStories several times, and it always seems to abort around 29%:

```
python litgpt/pretrain.py \
  --data litgpt.data.TinyStories \
  --model_name tiny-llama-1.1b \
  --tokenizer_dir checkpoints/TinyLlama/TinyLlama-1.1B-intermediate-step-1431k-3T/
```

```
 29%|███████████████████████████████                                                                              | 28508/100000 [00:49<01:55, 616.94it/s]


 29%|███████████████████████████████▉                                                                             | 29330/100000 [00:51<03:45, 313.37it/s]Done.██████████████████████████████▌                                                                              | 28019/100000 [00:50<04:04, 294.57it/s]
Traceback (most recent call last):█                                                                               | 27532/100000 [00:50<03:54, 308.75it/s]
  File "litgpt/pretrain.py", line 381, in <module>                                                                | 29038/100000 [00:51<02:40, 440.86it/s]
    CLI(setup)
  File "/teamspace/studios/this_studio/lit-gpt/litgpt/utils.py", line 399, in CLI                                 | 28618/100000 [00:50<03:36, 329.98it/s]
    return CLI(*args, **kwargs)████                                                                               | 27578/100000 [00:50<03:37, 333.52it/s]
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/jsonargparse/_cli.py", line 96, in CLI  | 28467/100000 [00:50<03:56, 301.86it/s]
    return _run_component(components, cfg_init)                                                                   | 27765/100000 [00:50<03:42, 324.88it/s]
  File "/home/zeus/miniconda3/envs/cloudspace/lib/python3.8/site-packages/jsonargparse/_cli.py", line 193, in _run_component0000 [00:50<02:53, 419.04it/s]
    return component(**cfg)████████▏                                                                              | 27640/100000 [00:50<03:44, 321.76it/s]
  File "litgpt/pretrain.py", line 90, in setup                                                                    | 27986/100000 [00:50<03:34, 336.14it/s]
    main(fabric, devices, seed, resume, config, data, out_dir, tokenizer_dir, tokenizer, train, eval)             | 27917/100000 [00:50<03:47, 316.19it/s]
  File "litgpt/pretrain.py", line 137, in main                                                                    | 27984/100000 [00:50<03:30, 342.61it/s]
    train_dataloader, val_dataloader = get_dataloaders(fabric, data, tokenizer, train, model.max_seq_length)      | 27595/100000 [00:50<02:55, 412.92it/s]
  File "litgpt/pretrain.py", line 319, in get_dataloaders                                                         | 27685/100000 [00:50<03:49, 314.98it/s]
    data.setup()██████████████████▉                                                                               | 27450/100000 [00:50<03:42, 326.29it/s]
  File "/teamspace/studios/this_studio/lit-gpt/litgpt/data/tinystories.py", line 57, in setup                     | 28620/100000 [00:50<02:43, 437.59it/s]
    assert len(shard_filenames) > 0, f"No bin files found in {bin_dir}"                                           | 27474/100000 [00:50<03:05, 391.62it/s]
AssertionError: No bin files found in data/TinyStories_all_data                                                   | 27420/100000 [00:50<04:16, 283.08it/s]
⚡ upd-prints ~/lit-gpt python litgpt/pretrain.py --data litgpt.data.TinyStories
```

For some reason, I can't neatly copy the terminal output because of the progress bars. It does show "done" every time though:

Screenshot 2024-03-11 at 4 40 44 PM

### PR DESCRIPTION
**Pull request by carmocca**
Fixes #995 
Fixes #1085

The tokenization is slightly different, this would be a bug in Karpathy's implementation.

This implementation no longer requires a ton of memory per worker

### CODE REVIEW

**Review by lantiga — APPROVED** [review]
Looks great

**Comment by lantiga** [comment]
Failures are legit though

**Comment by carmocca** [comment]
Yes, the tests need https://github.com/Lightning-AI/litdata/pull/77 to land and be released first

File: litgpt/data/tinystories.py [hunk 1]
```diff
@@ -27,155 +24,117 @@ class TinyStories(DataModule):
...
+    """The number of workers to use for the dataloaders."""
 
     tokenizer: Optional[Tokenizer] = field(default=None, init=False, repr=False)
     batch_size: int = field(default=1, init=False, repr=False)
     max_seq_length: int = field(default=-1, init=False, repr=False)
-    train_dataset: Optional[torch.utils.data.Dataset] = field(default=None, init=False, repr=False)
-    test_dataset: Optional[torch.utils.data.Dataset] = field(default=None, init=False, repr=False)
+
+    def __post_init__(self) -> None:
+        # Could be a remote path (s3://) or a local path
```
Comment (awaelchli) [inline litgpt/data/tinystories.py:40]: Do you want this dataset to support S3 paths?

**Review by awaelchli — APPROVED** [review]

File: litgpt/data/tinystories.py (hunk 1 above)
Comment (carmocca) [reply litgpt/data/tinystories.py:40]: If it's straightforward, sure. But it's not a priority, this is mainly for debugging and testing

**Pull request merged**